lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+    return;
+  }
+
+  // Register the dialects of the other container in this container.
+  Context->appendDialectRegistry(Container.Context->getDialectRegistry());
+
+  // Clone the other container's module into this container's context.
+  // This module is automatically erased at the end of scope.
+  OwningModuleRef TemporaryModule = cloneModuleInto(*Container.Module,
+                                                    *Context);
+
+  mlir::Block &Block = getModuleBlock(*Module);
+  visit(*TemporaryModule, [&](mlir::SymbolOpInterface Symbol) {
+    // Erase an existing symbol with the same name, if one exists.
+    if (auto S = mlir::SymbolTable::lookupSymbolIn(*Module, Symbol.getName()))
+      S->erase();
+
+    // Move each new symbol from the temporary module to the container's module.
+    Symbol->remove();
+    Block.push_back(Symbol);
+  });
+
+  // Assume that at least some symbols were copied over and always prune.
+  pruneSymbols(*Module);
+}
+
+pipeline::TargetsList MLIRContainer::enumerate() const {
+  pipeline::TargetsList::List List;
+
+  visit(Module.get(), [&](mlir::FunctionOpInterface F) {
+    if (F.isExternal())
+      return;
+
+    if (const auto MaybeTargetName = getFunctionName(F))
+      List.push_back(makeTarget(*MaybeTargetName));
+  });
+  llvm::sort(List);
+
+  return pipeline::TargetsList(std::move(List));
+}
+
+bool MLIRContainer::remove(const pipeline::TargetsList &List) {
+  if (getModuleBlock(*Module).empty())
+    return false;
+
+  bool RemovedSome = false;
+  visit(*Module, [&](mlir::FunctionOpInterface F) {
+    if (F.isExternal())
+      return;
+
+    const auto MaybeFunctionName = getFunctionName(F);
+    if (not MaybeFunctionName)
+      return;
+
+    makeExternal(F);
+    RemovedSome = true;
+  });
+
+  if (RemovedSome) {
+    // If any functions were removed, prune symbols and garbage collect types by
+    // cloning the module into a new context.
+
+    pruneSymbols(*Module);
+
+    auto NewContext = makeContext();
+    auto NewModule = cloneModuleInto(*Module, *NewContext);
+
+    Module = std::move(NewModule);
+    Context = std::move(NewContext);
+  }
+
+  return RemovedSome;
+}
+
+void MLIRContainer::clear() {
+  auto NewContext = makeContext();
+
+  Module = mlir::ModuleOp::create(mlir::UnknownLoc::get(NewContext.get()));
+  Context = std::move(NewContext);
+}
+
+llvm::Error MLIRContainer::serialize(llvm::raw_ostream &OS) const {
+  mlir::writeBytecodeToFile(*Module, OS);
+  return llvm::Error::success();
+}
+
+llvm::Error MLIRContainer::deserialize(const llvm::MemoryBuffer &Buffer) {
+  auto NewContext = makeContext();
+
+  const mlir::ParserConfig Config(NewContext.get());
+  OwningModuleRef
+    NewModule = mlir::parseSourceString<mlir::ModuleOp>(Buffer.getBuffer(),
+                                                        Config);
+
+  if (not NewModule)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Cannot load MLIR/LLVM module.");
```

Comment:
Why `/LLVM`?

Reply:
Originally this was just going to be a container for MLIR in the LLVM dialect, because I thought the different dialects would require very different code, but that's no longer the case. I'll fix it.